themes/bootstrap3/templates/search/searchbox.phtml

```diff
@@ -67,12 +68,41 @@
     <?php if (!empty($tabs)): ?></div><?php endif; ?>
   </div>
 <?php else: ?>
+  <div class="keyboard-box">
+    <div class="simple-keyboard"></div>
+  </div>
   <form id="searchForm" class="searchForm navbar-form navbar-left flip" method="get" action="<?=$this->url($basicSearch)?>" name="searchForm" autocomplete="off">
     <?= $this->context($this)->renderInContext('search/searchTabs', ['searchTabs' => $tabConfig['tabs'], 'showCounts' => $tabConfig['showCounts']]); ?>
     <?php $placeholder = $this->searchbox()->getPlaceholderText($tabConfig['selected']['id'] ?? null); ?>
-    <div class="searchForm-query">
+    <div class="searchForm-query input-group">
       <input id="searchForm_lookfor" class="searchForm_lookfor form-control search-query<?php if ($this->searchbox()->autocompleteEnabled($this->searchClassId)):?> autocomplete searcher:<?=$this->escapeHtmlAttr($this->searchClassId) ?><?=$this->searchbox()->autocompleteAutoSubmit($this->searchClassId) ? ' ac-auto-submit' : '' ?><?php endif ?>" type="search" name="lookfor" value="<?=$this->escapeHtmlAttr($this->lookfor)?>"<?php if ($placeholder): ?> placeholder="<?=$this->transEscAttr($placeholder) ?>"<?php endif ?> aria-label="<?=$this->transEscAttr('search_terms')?>">
-      <button id="searchForm-reset" class="searchForm-reset" type="reset" tabindex="-1" aria-label="<?=$this->transEsc('searchform_reset_button')?>"><?=$this->icon('ui-reset-search');?></button>
+      <div id="searchForm_controls">
+        <button id="searchForm-reset" class="searchForm-reset" type="button" tabindex="-1" aria-label="<?=$this->transEsc('searchform_reset_button')?>"><?=$this->icon('ui-reset-search');?></button>
+        <?php if (!empty($keyboardLayouts)): ?>
+          <?php
+          $this->headScript()->appendFile('vendor/js.cookie.js');
+          $this->headScript()->appendFile('vendor/simple-keyboard/index.js');
+          $this->headScript()->appendFile('vendor/simple-keyboard-layouts/index.js');
+          $this->headLink()->appendStylesheet('vendor/simple-keyboard/index.css');
+          ?>
+          <div class="keyboard-selection dropdown">
+            <button class="btn btn-primary dropdown-toggle" type="button" id="keyboard-selection-button" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">
+              <?=$this->icon('keyboard-o') ?>
+              <?=$this->icon('caret-down') ?>
+            </button>
+            <ul class="dropdown-menu" aria-labelledby="keyboard-selection-button">
+              <li>
+                <a class="keyboard-selection-item" href="#" data-value="none"><?= $this->transEsc('None') ?></a>
+              </li>
+              <?php foreach ($keyboardLayouts as $keyboardLayout): ?>
+                <li>
+                  <a class="keyboard-selection-item" href="#" data-value="<?= $keyboardLayout ?>"><?= $this->transEsc($keyboardLayout) ?></a>
```

Comment:
Should we create a text domain for these keyboard layouts so we can translate them without adding strings to the primary language files? That is, create `languages/KeyboardLayout/en.ini` with mappings, and then use `$this->transEsc("KeyboardLayout::$keyboardLayout")` here.

Reply:
I'm not sure about this. The layouts are names of languages and are often identical to the translation keys of the language selection except for the case of the first letter. Wouldn't it be better to use the same translations there instead of duplicating most of it?

Reply:
We actually don't have language translations yet outside of native.ini, which serves a somewhat different purpose. Broader translations are the subject of #413, but it hasn't been resolved since no one has had time to reach consensus on language representation in the index and to build the necessary i18n files. I agree that these efforts could be related. Perhaps this is an opportunity to start something.

Reply:
(And when I say this is different from native.ini, my thought is that the language selection control helps people find their native language -- i.e. I can spot English in the list, a German- speaker will recognize Deutsche, etc. -- but I think the virtual keyboard labels should be translated to the active language. I might need a Russian keyboard even if I can't speak Russian but need to enter some special characters, but I'd have difficulty knowing which Cyrillic word means Russian.

Reply:
I'm open to other arguments, though; maybe my first impression doesn't reflect the real use cases!

Reply:
I think you are right about handling language selection translations different than in the keyboard selection.  

I shared an idea in #413 how this could be done.

Reply:
Some work is moving forward now in #3200. Do you think we should focus on finishing that before wrapping this up, or should we find an interim solution here and add a TODO there to update things to handle keyboard layouts when possible?

Reply:
I think we can finish this now. Other PRs might introduce conflicts again in the future. And I don't think that having translations for the layouts is that critical. Even if they are not included in 10.0 if anyone wants to use this feature they can add the translations in their local configs.

Reply:
I've added a TODO item on #3200 to set up translation for keyboard layout names so that we don't lose track of this task.